src/ripple/app/tx/impl/InvariantCheck.cpp

```diff
@@ -0,0 +1,161 @@
+
+#include <ripple/app/tx/impl/InvariantCheck.h>
+#include <ripple/basics/Log.h>
+
+namespace ripple {
+
+void
+XRPNotCreated::visitEntry(
+    uint256 const&,
+    bool isDelete,
+    std::shared_ptr <SLE const> const& before,
+    std::shared_ptr <SLE const> const& after)
+{
+    if(before)
+    {
+        switch (before->getType())
+        {
+        case ltACCOUNT_ROOT:
+            drops_ -= (*before)[sfBalance].xrp().drops();
+            break;
+        case ltPAYCHAN:
+            drops_ -= ((*before)[sfAmount] - (*before)[sfBalance]).xrp().drops();
+            break;
+        case ltESCROW:
+            drops_ -= (*before)[sfAmount].xrp().drops();
+            break;
+        default:
+            break;
+        }
+    }
+
+    if(after)
+    {
+        switch (after->getType())
+        {
+        case ltACCOUNT_ROOT:
+            drops_ += (*after)[sfBalance].xrp().drops();
+            break;
+        case ltPAYCHAN:
+            if (! isDelete)
+                drops_ += ((*after)[sfAmount] - (*after)[sfBalance]).xrp().drops();
+            break;
+        case ltESCROW:
+            if (! isDelete)
+                drops_ += (*after)[sfAmount].xrp().drops();
+            break;
+        default:
+            break;
+        }
+    }
+}
+
+bool
+XRPNotCreated::finalize(STTx const& tx, TER /*tec*/, beast::Journal const& j)
+{
+    auto fee = tx.getFieldAmount(sfFee).xrp().drops();
+    if(-1*fee <= drops_ && drops_ <= 0)
+        return true;
+
+    JLOG(j.fatal()) << "Invariant failed: XRP net change was " << drops_ <<
+        " on a fee of " << fee;
+    return false;
+}
+
+//------------------------------------------------------------------------------
+
+void
+AccountRootsNotDeleted::visitEntry(
+    uint256 const&,
+    bool isDelete,
+    std::shared_ptr <SLE const> const& before,
+    std::shared_ptr <SLE const> const& /*after*/)
+{
+    if (isDelete && before && before->getType() == ltACCOUNT_ROOT)
+        accountDeleted_ = true;
+}
+
+bool
+AccountRootsNotDeleted::finalize(STTx const&, TER, beast::Journal const& j)
+{
+    if (! accountDeleted_)
+        return true;
+
+    JLOG(j.fatal()) << "Invariant failed: an account root was deleted";
+    return false;
+}
+
+//------------------------------------------------------------------------------
+
+void
+LedgerEntryTypesMatch::visitEntry(
+    uint256 const&,
+    bool,
+    std::shared_ptr <SLE const> const& before,
+    std::shared_ptr <SLE const> const& after)
+{
+    if (before && after && before->getType() != after->getType())
+        typeMismatch_ = true;
+
+    if (after)
+    {
+        switch (after->getType())
+        {
+        case ltACCOUNT_ROOT:
+        case ltDIR_NODE:
+        case ltRIPPLE_STATE:
+        case ltTICKET:
+        case ltSIGNER_LIST:
+        case ltOFFER:
+        case ltLEDGER_HASHES:
+        case ltAMENDMENTS:
+        case ltFEE_SETTINGS:
+        case ltESCROW:
```

Comment:
I don't believe we're passing rules or the feature set into the invariants right now, so we would have to add that if we wanted to do this. I'm happy to look into it more if there is sufficient value...